Handle 'Cancelled' build status in the web console

spadgett opened the pull request:
Fixes #2144

Browse -> Builds

![openshift_management_console](https://cloud.githubusercontent.com/assets/1167259/7544390/9924e8f6-f59a-11e4-815d-cfaeaeb7e1b2.png)

Overview

![openshift_management_console](https://cloud.githubusercontent.com/assets/1167259/7544388/8e742f20-f59a-11e4-95f3-675e4daf0318.png)

spadgett commented:
@liggitt please review

liggitt commented on assets/app/views/_triggers.html:
```diff
@@ -3,9 +3,16 @@
     <span ng-switch="build.status" class="hide-ng-leave">
       <span ng-switch-when="Complete" class="fa fa-check text-success" aria-hidden="true" style="margin-right: 5px;"></span>
       <span ng-switch-when="Failed" class="fa fa-times text-danger" aria-hidden="true" style="margin-right: 5px;"></span>
+      <span ng-switch-when="Cancelled" class="fa fa-ban text-warning" aria-hidden="true"></span>
```
do we know all the states? Seems like we might want a hidden icon to be the default, and only show spinning for "Pending", "Running", etc

spadgett replied:
@liggitt Looks like these are the states. We should handle Error as well.

```
// Valid values for BuildPhase.
const (
	// BuildPhaseNew is automatically assigned to a newly created build.
	BuildPhaseNew BuildPhase = "New"

	// BuildPhasePending indicates that a pod name has been assigned and a build is
	// about to start running.
	BuildPhasePending BuildPhase = "Pending"

	// BuildPhaseRunning indicates that a pod has been created and a build is running.
	BuildPhaseRunning BuildPhase = "Running"

	// BuildPhaseComplete indicates that a build has been successful.
	BuildPhaseComplete BuildPhase = "Complete"

	// BuildPhaseFailed indicates that a build has executed and failed.
	BuildPhaseFailed BuildPhase = "Failed"

	// BuildPhaseError indicates that an error prevented the build from executing.
	BuildPhaseError BuildPhase = "Error"

	// BuildPhaseCancelled indicates that a running/pending build was stopped from executing.
	BuildPhaseCancelled BuildPhase = "Cancelled"
)
```

spadgett replied:
We might also add tooltip help for status to Browse -> Builds.

liggitt commented on assets/app/views/_triggers.html:
```diff
@@ -3,9 +3,16 @@
-    <span>A build of {{build.metadata.labels.buildconfig}} is {{build.status | lowercase}}.</span>
-    <span ng-if="build.status != 'Complete' && build.status != 'Failed' && trigger.imageChangeParams.automatic">  A new deployment will be created automatically once the build completes.</span>
+    <span>
+      A build of {{build.metadata.labels.buildconfig}}
```
probably need another switch
Failed: A build of ... failed
Cancelled: A build of ... was cancelled
Default: A build of ... is lower(status)

spadgett commented:
@liggitt PR updated when you have time

liggitt commented on assets/app/views/_triggers.html:
```diff
@@ -1,11 +1,25 @@
-    <span ng-if="build.status != 'Complete' && build.status != 'Failed' && trigger.imageChangeParams.automatic">  A new deployment will be created automatically once the build completes.</span>
+    <!-- Message -->
+    <span ng-switch="build.status" class="hide-ng-leave">
+      <span ng-switch-when="Complete">A build of {{build.metadata.labels.buildconfig}} completed.</span>
```
factor out `A build of {{build.metadata.labels.buildconfig}}` so we don't have to repeat it. Error can say "... encountered an error."

liggitt commented on assets/app/views/_triggers.html:
```diff
@@ -1,11 +1,25 @@
+        A build of {{build.metadata.labels.buildconfig}} is {{build.status | lowercase}}.
+      </span>
+    </span>
+    <span ng-if="build.status != 'Complete' && build.status != 'Failed' && build.status != 'Cancelled' && trigger.imageChangeParams.automatic">
```
this can nest inside the `ng-switch-default` case above and drop all the `build.status != ...` checks

spadgett commented:
@liggitt I noticed a problem where Cancelled builds were not cleared from the Overview page after 5 minutes. I've fixed that now and updated the PR with your suggestions.

liggitt commented on assets/app/views/_triggers.html:
```diff
@@ -1,11 +1,25 @@
 <div class="builds" ng-repeat="trigger in triggers">
-  <div ng-repeat="build in trigger.builds" ng-if="(build.status != 'Complete' && build.status != 'Failed') || (build.metadata.creationTimestamp | ageLessThan : 5 : 'minutes')" class="build animate-repeat osc-object" kind="Build" resource="build">
+  <div ng-repeat="build in trigger.builds" ng-if="build.status == 'New' || build.status == 'Pending' || build.status == 'Running' || (build.metadata.creationTimestamp | ageLessThan : 5 : 'minutes')" class="build animate-repeat osc-object" kind="Build" resource="build">
```
Probably worth separating the check if the build is running into a filter ("isBuildInProgress", etc) that checks new, pending, running

liggitt replied:
Fine to do that in a separate PR

liggitt commented:
[merge]
